### MovieLens Case Study

The GroupLens Research Project is a research group in the Department of Computer Science and Engineering at the University of Minnesota. The data is widely used for collaborative filtering and other filtering solutions. However, we will be using this data to act as a means to demonstrate our skill in using Python to “play” with data.

### Datasets Information:

- **ratings.csv:** It contains information on ratings given by the users to a particular movie. Columns: user id, movie id, rating, timestamp

- **movie.csv:** The file contains information related to the movies and their genre. Columns: movie id, movie title, release date, unknown, Action, Adventure, Animation, Children’s, Comedy, Crime, Documentary, Drama, Fantasy, Film-Noir, Horror, Musical, Mystery, Romance, Sci-Fi, Thriller, War, Western

- **user.csv:** It contains information about the users who have rated the movies.. Columns: user id, age, gender, occupation, zip code

### Objective:

`To extract insights from the dataset`

### Learning Outcomes:
`Use of Pandas Functions - shape, describe, groupby, merge etc.`


#### Domain 
`Internet and Entertainment`

**Note that the case study will need you to apply the concepts of groupby and merging extensively.**

### 1. Import the necessary packages

In [1]:
import pandas as pd
import numpy as np

### 2. Read all the three datasets

In [2]:
# Reading datasets by using read_csv from pandas package
ratings = pd.read_csv("./data/ratings.csv")
movie = pd.read_csv("./data/movie.csv")
user = pd.read_csv("./data/user.csv")

### 3. View the first 5 rows of all the datasets.
`Note that you will need to do it for all three datasets separately`

In [3]:
ratings.head(5)

user id  movie id  rating  timestamp
0      196       242       3  881250949
1      186       302       3  891717742
2       22       377       1  878887116
3      244        51       2  880606923
4      166       346       1  886397596

In [4]:
movie.head(5)

movie id  movie title release date  Action  Adventure  Animation  \
0         1   Toy Story      1-Jan-95       0          0          1   
1         2   GoldenEye      1-Jan-95       1          1          0   
2         3  Four Rooms      1-Jan-95       0          0          0   
3         4  Get Shorty      1-Jan-95       1          0          0   
4         5     Copycat      1-Jan-95       0          0          0   

   Childrens  Comedy  Crime  Documentary  ...  Fantasy  Film-Noir  Horror  \
0          1       1      0            0  ...        0          0       0   
1          0       0      0            0  ...        0          0       0   
2          0       0      0            0  ...        0          0       0   
3          0       1      0            0  ...        0          0       0   
4          0       0      1            0  ...        0          0       0   

   Musical  Mystery  Romance  Sci-Fi  Thriller  War  Western  
0        0        0        0       0         0    0        0  
1        0        0        0       0         1    0        0  
2        0        0        0       0         1    0        0  
3        0        0        0       0         0    0        0  
4        0        0        0       0         1    0        0  

[5 rows x 21 columns]

In [6]:
user.head(10)

user id  age gender     occupation zip code
0        1   24      M     technician    85711
1        2   53      F          other    94043
2        3   23      M         writer    32067
3        4   24      M     technician    43537
4        5   33      F          other    15213
5        6   42      M      executive    98101
6        7   57      M  administrator    91344
7        8   36      M  administrator    05201
8        9   29      M        student    01002
9       10   53      M         lawyer    90703

### 4. Understand the shape of all the datasets.
`Note that you will need to do it for all the three datasets seperately`

In [7]:
# ratings
ratings.shape

(100000, 4)

 **Observation:** There are 100000 rows and 4 columns in the ratings dataset

In [8]:
# user
user.shape

(943, 5)

 **Observation:** There are 943 rows and 5 columns in the user dataset

In [9]:
# movie
movie.shape

(1680, 21)

 **Observation:** There are 1680 rows and 21 columns in the movie dataset

### 5. Check the data types of the columns for all the datasets.
 `Note that you will need to do it for all three datasets separately`

In [10]:
# ratings
# We use dataframe.dtypes to get the data types of each column
ratings.dtypes 

user id      int64
movie id     int64
rating       int64
timestamp    int64
dtype: object

 **Observation:** All columns have integer data type 

In [11]:
# user
user.dtypes

user id        int64
age            int64
gender        object
occupation    object
zip code      object
dtype: object

 **Observations:**
 1. user id and age columns are of integer data types 
 2. gender, occupation and zip code columns are of string data type

In [12]:
# movie
movie.dtypes

movie id         int64
movie title     object
release date    object
Action           int64
Adventure        int64
Animation        int64
Childrens        int64
Comedy           int64
Crime            int64
Documentary      int64
Drama            int64
Fantasy          int64
Film-Noir        int64
Horror           int64
Musical          int64
Mystery          int64
Romance          int64
Sci-Fi           int64
Thriller         int64
War              int64
Western          int64
dtype: object

**Observation:**
1. movie title and release date are of string data type
2. movie id and all genres are of interger data type

### 6. Give a statistical summary for all the datasets.
`Note that you will need to do it for all three datasets separately`

In [13]:
# ratings
ratings.describe()

user id       movie id         rating     timestamp
count  100000.00000  100000.000000  100000.000000  1.000000e+05
mean      462.48475     425.530130       3.529860  8.835289e+08
std       266.61442     330.798356       1.125674  5.343856e+06
min         1.00000       1.000000       1.000000  8.747247e+08
25%       254.00000     175.000000       3.000000  8.794487e+08
50%       447.00000     322.000000       4.000000  8.828269e+08
75%       682.00000     631.000000       4.000000  8.882600e+08
max       943.00000    1682.000000       5.000000  8.932866e+08

**Observation:** The mean and Median of user ratings are 3.53 & 4.00 respectively

In [14]:
# user
user.describe()

user id         age
count  943.000000  943.000000
mean   472.000000   34.051962
std    272.364951   12.192740
min      1.000000    7.000000
25%    236.500000   25.000000
50%    472.000000   31.000000
75%    707.500000   43.000000
max    943.000000   73.000000

**Observation:** The average age of all the users is 34 years while the range lies between 7 to 73 years.

In [15]:
# movie
movie.describe()

movie id       Action    Adventure    Animation    Childrens  \
count  1680.000000  1680.000000  1680.000000  1680.000000  1680.000000   
mean    841.525595     0.149405     0.080357     0.025000     0.072619   
std     485.609591     0.356593     0.271926     0.156171     0.259587   
min       1.000000     0.000000     0.000000     0.000000     0.000000   
25%     421.750000     0.000000     0.000000     0.000000     0.000000   
50%     841.500000     0.000000     0.000000     0.000000     0.000000   
75%    1261.250000     0.000000     0.000000     0.000000     0.000000   
max    1682.000000     1.000000     1.000000     1.000000     1.000000   

            Comedy        Crime  Documentary        Drama      Fantasy  \
count  1680.000000  1680.000000  1680.000000  1680.000000  1680.000000   
mean      0.300595     0.064881     0.029762     0.431548     0.013095   
std       0.458653     0.246389     0.169980     0.495440     0.113717   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       1.000000     0.000000     0.000000     1.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

         Film-Noir       Horror      Musical      Mystery      Romance  \
count  1680.000000  1680.000000  1680.000000  1680.000000  1680.000000   
mean      0.014286     0.054762     0.033333     0.036310     0.147024   
std       0.118701     0.227583     0.179559     0.187115     0.354235   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       0.000000     0.000000     0.000000     0.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

            Sci-Fi     Thriller          War      Western  
count  1680.000000  1680.000000  1680.000000  1680.000000  
mean      0.060119     0.149405     0.042262     0.016071  
std       0.237778     0.356593     0.201246     0.125788  
min       0.000000     0.000000     0.000000     0.000000  
25%       0.000000     0.000000     0.000000     0.000000  
50%       0.000000     0.000000     0.000000     0.000000  
75%       0.000000     0.000000     0.000000     0.000000  
max       1.000000     1.000000     1.000000     1.000000

**Observation:** The genres should be in categorical format and not in the numeric because it is of binary class

### The statistical summary of the data will give us an overview of the data by providing key statistics of the different columns in the data.

### 7. Find the number of movies per genre using the movie data

In [16]:
# Getting all the column names
movie.columns

Index(['movie id', 'movie title', 'release date', 'Action', 'Adventure',
       'Animation', 'Childrens', 'Comedy', 'Crime', 'Documentary', 'Drama',
       'Fantasy', 'Film-Noir', 'Horror', 'Musical', 'Mystery', 'Romance',
       'Sci-Fi', 'Thriller', 'War', 'Western'],
      dtype='object')

In [17]:
type(movie)

pandas.core.frame.DataFrame

In [19]:
type(movie.columns)

pandas.core.indexes.base.Index

In [22]:
list_of_genres = [ 'Action',
       'Adventure', 'Animation', 'Childrens', 'Comedy', 'Crime', 'Documentary',
       'Drama', 'Fantasy', 'Film-Noir', 'Horror', 'Musical', 'Mystery',
       'Romance', 'Sci-Fi', 'Thriller', 'War', 'Western']

In [24]:
movie[list_of_genres].sum()

Action         251
Adventure      135
Animation       42
Childrens      122
Comedy         505
Crime          109
Documentary     50
Drama          725
Fantasy         22
Film-Noir       24
Horror          92
Musical         56
Mystery         61
Romance        247
Sci-Fi         101
Thriller       251
War             71
Western         27
dtype: int64

In [26]:
type(movie[list_of_genres].sum())

## Subsetting

pandas.core.series.Series

In [20]:
# Taking all the genre columns and finding the sum for every column
movie[[ 'Action',
       'Adventure', 'Animation', 'Childrens', 'Comedy', 'Crime', 'Documentary',
       'Drama', 'Fantasy', 'Film-Noir', 'Horror', 'Musical', 'Mystery',
       'Romance', 'Sci-Fi', 'Thriller', 'War', 'Western']].sum() 

Action         251
Adventure      135
Animation       42
Childrens      122
Comedy         505
Crime          109
Documentary     50
Drama          725
Fantasy         22
Film-Noir       24
Horror          92
Musical         56
Mystery         61
Romance        247
Sci-Fi         101
Thriller       251
War             71
Western         27
dtype: int64

In [27]:
# Alternatively, we can also loc function
movie.loc[:,'Action':'Western'].sum()

Action         251
Adventure      135
Animation       42
Childrens      122
Comedy         505
Crime          109
Documentary     50
Drama          725
Fantasy         22
Film-Noir       24
Horror          92
Musical         56
Mystery         61
Romance        247
Sci-Fi         101
Thriller       251
War             71
Western         27
dtype: int64

In [28]:
number = movie.loc[:,'Action':'Western'].sum()

In [29]:
type(number)

pandas.core.series.Series

In [30]:
# Sorting the movies across genres
number = movie.loc[:,'Action':'Western'].sum()
number.sort_values(ascending = False)

Drama          725
Comedy         505
Action         251
Thriller       251
Romance        247
Adventure      135
Childrens      122
Crime          109
Sci-Fi         101
Horror          92
War             71
Mystery         61
Musical         56
Documentary     50
Animation       42
Western         27
Film-Noir       24
Fantasy         22
dtype: int64

**Observations:**
1. Drama and Comedy are the most common movie genre.
2. Clearly, there are some movies that have more than one genre.

### Applying functions on specific columns of the data will help us in gaining specific insights about the data.

### 8. Find the movies that have more than one genre
`Hint: use sum on the axis = 1`

In [31]:
# Checking column names
movie.columns

Index(['movie id', 'movie title', 'release date', 'Action', 'Adventure',
       'Animation', 'Childrens', 'Comedy', 'Crime', 'Documentary', 'Drama',
       'Fantasy', 'Film-Noir', 'Horror', 'Musical', 'Mystery', 'Romance',
       'Sci-Fi', 'Thriller', 'War', 'Western'],
      dtype='object')

In [32]:
# we create a new dataframe using two columns of the movie dataframe
new_movie = movie[['movie id', 'movie title']].copy()

In [33]:
new_movie.head()

movie id  movie title
0         1   Toy Story 
1         2   GoldenEye 
2         3  Four Rooms 
3         4  Get Shorty 
4         5     Copycat

In [34]:
new_movie["Number of Genres"] = movie.loc[:, 'Action':'Western'].sum(axis=1)

In [35]:
new_movie

movie id                          movie title  Number of Genres
0            1                           Toy Story                  3
1            2                           GoldenEye                  3
2            3                          Four Rooms                  1
3            4                          Get Shorty                  3
4            5                             Copycat                  3
...        ...                                  ...               ...
1675      1678                          Mat' i syn                  1
1676      1679                           B. Monkey                  2
1677      1680                       Sliding Doors                  2
1678      1681                        You So Crazy                  1
1679      1682  Scream of Stone (Schrei aus Stein)                  1

[1680 rows x 3 columns]

In [ ]:
columns(movie)

In [ ]:
# df['new_column_name'] = pd.some_pandas_command

In [39]:
# Filtering movies that have more than 1 genres
new_movie[new_movie['Number of Genres'] > 1]

movie id         movie title  Number of Genres
0            1          Toy Story                  3
1            2          GoldenEye                  3
3            4         Get Shorty                  3
4            5            Copycat                  3
6            7     Twelve Monkeys                  2
...        ...                 ...               ...
1666      1669  MURDER and murder                  3
1667      1670            Tainted                  2
1670      1673             Mirage                  2
1676      1679          B. Monkey                  2
1677      1680      Sliding Doors                  2

[849 rows x 3 columns]

**Observation:** 849 movies have more than one genre.

### Creating new columns using existing columns in the dataframe requires domain knowledge, and can help us to understand our data better.

### 9. Find the top 25 movies according to average ratings such that each movie has number of ratings more than 100

Hint : 

1. First find the movies that have more than 100 ratings(use groupby and count). Extract the movie id in a list.
2. Find the average rating of all the movies and sort them in the descending order. 
3. Use isin(list obtained from 1) to filter out the movies which have more than 100 ratings.
4. You will have to use the .merge() function to get the movie titles.

Note: This question will need you to research about groupby and apply your findings. You can find more on groupby on https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html.

In [40]:
# Merging ratings dataset with movie dataset
df_merge = movie.merge(ratings, on = 'movie id', how = 'inner')
df_merge.head()

movie id movie title release date  Action  Adventure  Animation  Childrens  \
0         1  Toy Story      1-Jan-95       0          0          1          1   
1         1  Toy Story      1-Jan-95       0          0          1          1   
2         1  Toy Story      1-Jan-95       0          0          1          1   
3         1  Toy Story      1-Jan-95       0          0          1          1   
4         1  Toy Story      1-Jan-95       0          0          1          1   

   Comedy  Crime  Documentary  ...  Musical  Mystery  Romance  Sci-Fi  \
0       1      0            0  ...        0        0        0       0   
1       1      0            0  ...        0        0        0       0   
2       1      0            0  ...        0        0        0       0   
3       1      0            0  ...        0        0        0       0   
4       1      0            0  ...        0        0        0       0   

   Thriller  War  Western  user id  rating  timestamp  
0         0    0        0      308       4  887736532  
1         0    0        0      287       5  875334088  
2         0    0        0      148       4  877019411  
3         0    0        0      280       4  891700426  
4         0    0        0       66       3  883601324  

[5 rows x 24 columns]

In [41]:
# Checking the dimensions of the merged dataframe
df_merge.shape

(99990, 24)

In [46]:
type(df_merge.groupby(['movie title'])['rating'].count().reset_index())

pandas.core.frame.DataFrame

In [42]:
# Finding the count of ratings for each movie using groupby() and count()
# reset_index() is used to shift movie title from being the dataframe’s (movie_count’s) index to 
# being just a normal column 
movie_count = df_merge.groupby(['movie title'])['rating'].count().reset_index()
movie_count.head()
# movie_count = df_merge.groupby(['movie title'])['rating'].count()
# type(movie_count)

movie title  rating
0  'Til There Was You        9
1               1-900        5
2      101 Dalmatians      109
3        12 Angry Men      125
4                  187      41

In [47]:
# Extracting the movie titles that have more than 100 ratings 
movie_100 = movie_count[movie_count['rating']>100]['movie title']
movie_100.head()

2            101 Dalmatians 
3              12 Angry Men 
7     2001: A Space Odyssey 
15           Absolute Power 
16               Abyss, The 
Name: movie title, dtype: object

In [48]:
movie_count[movie_count['rating']>100]['movie title'].reset_index()


index                             movie title
0        2                         101 Dalmatians 
1        3                           12 Angry Men 
2        7                  2001: A Space Odyssey 
3       15                         Absolute Power 
4       16                             Abyss, The 
..     ...                                     ...
329   1627  Willy Wonka and the Chocolate Factory 
330   1638                      Wizard of Oz, The 
331   1646                    Wrong Trousers, The 
332   1651                     Young Frankenstein 
333   1652                             Young Guns 

[334 rows x 2 columns]

In [50]:
# Finding average ratings for each movie and sorting them out in descending order
# using groupby() and sort_values() on merged data frame
avg_rating = df_merge.groupby(['movie title'])['rating'].mean().sort_values(ascending=False).reset_index()
avg_rating

movie title  rating
0                  Saint of Fort Washington, The      5.0
1     Entertaining Angels: The Dorothy Day Story      5.0
2                                  Aiqing wansui      5.0
3                             Santa with Muscles      5.0
4                                    Prefontaine      5.0
...                                           ...     ...
1652         Tigrero: A Film That Was Never Made      1.0
1653                               Low Life, The      1.0
1654                              Pharaoh's Army      1.0
1655                        To Cross the Rubicon      1.0
1656                                    Power 98      1.0

[1657 rows x 2 columns]

In [53]:
# Extracting movie titles that have more than 100 ratings using movie titles in movie_100 and isin() function
# Displaying top 25 rows only
avg_rating[avg_rating['movie title'].isin(movie_100)]

movie title    rating
15               Close Shave, A   4.491071
16             Schindler's List   4.466443
17          Wrong Trousers, The   4.466102
18                   Casablanca   4.456790
20    Shawshank Redemption, The   4.445230
...                          ...       ...
1261                      Spawn   2.615385
1277              Event Horizon   2.574803
1291                      Crash   2.546875
1357              Jungle2Jungle   2.439394
1392             Cable Guy, The   2.339623

[334 rows x 2 columns]

### We have just learned how to use groupby() to  group the various categories in the data, and how to use sort_values() to sort the data.

### 10. Find the relationship between user demographics and movie ratings

#### 10.1 Find the mean rating for every occupation 

1. Merge the user dataset with movie and ratings(already merged : df_merge) dataset
2. Apply groupby on occupation

In [54]:
# Merging user dataset with movie and ratings(already merged : df_merge) dataset
df_merge_all = df_merge.merge(user, on = 'user id', how = 'inner')

In [55]:
df_merge_all.shape

(99990, 28)

In [57]:
df_merge_all.head()

movie id      movie title release date  Action  Adventure  Animation  \
0         1       Toy Story      1-Jan-95       0          0          1   
1         4      Get Shorty      1-Jan-95       1          0          0   
2         5         Copycat      1-Jan-95       0          0          0   
3         7  Twelve Monkeys      1-Jan-95       0          0          0   
4         8            Babe      1-Jan-95       0          0          0   

   Childrens  Comedy  Crime  Documentary  ...  Thriller  War  Western  \
0          1       1      0            0  ...         0    0        0   
1          0       1      0            0  ...         0    0        0   
2          0       0      1            0  ...         1    0        0   
3          0       0      0            0  ...         0    0        0   
4          1       1      0            0  ...         0    0        0   

   user id  rating  timestamp  age  gender  occupation  zip code  
0      308       4  887736532   60       M     retired     95076  
1      308       5  887737890   60       M     retired     95076  
2      308       4  887739608   60       M     retired     95076  
3      308       4  887738847   60       M     retired     95076  
4      308       5  887736696   60       M     retired     95076  

[5 rows x 28 columns]

In [56]:
# Group by occupation and aggregate with mean
df_merge_all.groupby('occupation').rating.mean()

occupation
administrator    3.635646
artist           3.653380
doctor           3.688889
educator         3.670692
engineer         3.541473
entertainment    3.440783
executive        3.349794
healthcare       2.896220
homemaker        3.301003
lawyer           3.735316
librarian        3.560781
marketing        3.485641
none             3.777778
other            3.552335
programmer       3.568205
retired          3.466750
salesman         3.582944
scientist        3.611273
student          3.515167
technician       3.532097
writer           3.376152
Name: rating, dtype: float64

**Observation:**: The mean rating is low (2.89) for people working in healthcare.

#### 10.2 Find mean rating for each gender

In [58]:
df_merge_all.groupby('gender').rating.mean()

gender
F    3.531510
M    3.529333
Name: rating, dtype: float64

**Observation**: The ratings given by men and women are similar on average.

#### 10.3 Find the mean rating grouped by both occupation and gender
1. Put both the columns under groupby function and find the mean rating. 

In [59]:
# https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html

In [61]:
df_merge_all.groupby(['occupation','gender']).rating.mean().reset_index()

occupation gender    rating
0   administrator      F  3.781839
1   administrator      M  3.555233
2          artist      F  3.347065
3          artist      M  3.875841
4          doctor      M  3.688889
5        educator      F  3.699132
6        educator      M  3.660246
7        engineer      F  3.751724
8        engineer      M  3.537676
9   entertainment      F  3.448889
10  entertainment      M  3.439807
11      executive      F  3.773756
12      executive      M  3.320340
13     healthcare      F  2.736021
14     healthcare      M  3.639839
15      homemaker      F  3.278810
16      homemaker      M  3.500000
17         lawyer      F  3.623188
18         lawyer      M  3.741379
19      librarian      F  3.580070
20      librarian      M  3.537920
21      marketing      F  3.522624
22      marketing      M  3.474801
23           none      F  3.632877
24           none      M  3.876636
25          other      F  3.531114
26          other      M  3.563447
27     programmer      F  3.577566
28     programmer      M  3.567674
29        retired      F  3.239437
30        retired      M  3.477243
31       salesman      F  3.870206
32       salesman      M  3.394584
33      scientist      F  3.251799
34      scientist      M  3.637311
35        student      F  3.602879
36        student      M  3.484440
37     technician      F  3.268519
38     technician      M  3.540477
39         writer      F  3.663986
40         writer      M  3.180770

**Observations:** 

1. We had seen that the healthcare sector had a low mean rating. We now see that the low rating is driven by the female workers in the healthcare sector.

2. The male writers have given lower ratings compared to the female writers.

3. Only male doctors have given ratings. No female doctors have given ratings as per the data.

### Using groupby() on multiple columns of the data helps us analyze subsets of the data to gain insights.

### 11. Find the mean rating for each genre.

In [62]:
cols=df_merge_all.loc[:,'Action':'Western'].columns


In [63]:
cols

Index(['Action', 'Adventure', 'Animation', 'Childrens', 'Comedy', 'Crime',
       'Documentary', 'Drama', 'Fantasy', 'Film-Noir', 'Horror', 'Musical',
       'Mystery', 'Romance', 'Sci-Fi', 'Thriller', 'War', 'Western'],
      dtype='object')

In [64]:
# cols=df_merge_all.loc[:,'Action':'Western'].columns

for i in cols:
    print(i,':' , df_merge_all[df_merge_all[i]==1].rating.mean())

Action : 3.480245417953027
Adventure : 3.503526503308369
Animation : 3.5766990291262135
Childrens : 3.3532442216652742
Comedy : 3.3940734781442745
Crime : 3.6322780881440098
Documentary : 3.6728232189973613
Drama : 3.6873793708484772
Fantasy : 3.2152366863905324
Film-Noir : 3.9215233698788228
Horror : 3.2903893172841827
Musical : 3.521396851029471
Mystery : 3.63813155386082
Romance : 3.621704948358255
Sci-Fi : 3.5607227022780834
Thriller : 3.5090069495245064
War : 3.815811874866993
Western : 3.6132686084142396


**Observation**: War movies have the highest average rating, while Fantasy movies have the lowest average rating.

### 12. Divide the age column into four groups.

* Group 1. 0-20 => Teenager
* Group 2. 20-40 => Adult
* Group 3. 40-55 => Middle Age
* Group 4. 55+ => Elderly

This questions requires pandas cut() function.

Refer this link for more info: https://pandas.pydata.org/pandas-docs/version/0.23.4/generated/pandas.cut.html

#### Updated Link: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.cut.html

In [65]:
#create a new column with name age_group
df_merge_all['age_group']=pd.cut(df_merge_all.age, bins=(0,20,40,55,100),labels=('Teenager','Adult','Middle Age','Elderly'))

In [67]:
#print top 5 entries of the age and age_group columns
df_merge_all[['age','age_group']]

age age_group
0       60   Elderly
1       60   Elderly
2       60   Elderly
3       60   Elderly
4       60   Elderly
...    ...       ...
99985   19  Teenager
99986   19  Teenager
99987   19  Teenager
99988   19  Teenager
99989   19  Teenager

[99990 rows x 2 columns]

In [68]:
#print last 5 entries of the age and age_group columns
df_merge_all[['age','age_group']].tail()

age age_group
99985   19  Teenager
99986   19  Teenager
99987   19  Teenager
99988   19  Teenager
99989   19  Teenager

#### 12.1 Find the mean rating for each age group.

In [69]:
df_merge_all.groupby(['age_group']).rating.mean()

age_group
Teenager      3.548290
Adult         3.492513
Middle Age    3.606980
Elderly       3.651703
Name: rating, dtype: float64

**Observation:** Old people have given higher ratings that any other age group.

#### 12.2 Find the mean, median, and standard deviation of rating for each age group.

1. This can be done using groupby.agg(), where agg() stands for aggregate.
2. agg() can take more than 1 aggregate function simultaneously.

Refer to this link for more info :https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.core.groupby.DataFrameGroupBy.aggregate.html

In [70]:
df_merge_all.groupby(['age_group']).rating.agg(['mean','median','std'])

mean  median       std
age_group                             
Teenager    3.548290       4  1.154198
Adult       3.492513       4  1.138085
Middle Age  3.606980       4  1.083602
Elderly     3.651703       4  1.030762

#### 12.3 Which genre movies are common among Engineers?

In [72]:
df_merge_all.groupby('occupation').sum().loc[:,'Action':'Western'].loc['engineer'].reset_index()

index  engineer
0        Action      2277
1     Adventure      1299
2     Animation       363
3     Childrens       629
4        Comedy      2438
5         Crime       611
6   Documentary        51
7         Drama      3153
8       Fantasy       125
9     Film-Noir       145
10       Horror       414
11      Musical       441
12      Mystery       363
13      Romance      1467
14       Sci-Fi      1221
15     Thriller      1712
16          War       858
17      Western       180

In [73]:
df_merge_all.groupby('occupation').sum().loc[:,'Action':'Western']

Action  Adventure  Animation  Childrens  Comedy  Crime  \
occupation                                                              
administrator    1858       1020        203        462    2203    565   
artist            528        311        121        202     617    201   
doctor            110         57         22         28     168     41   
educator         1962       1128        263        593    2708    710   
engineer         2277       1299        363        629    2438    611   
entertainment     499        249         62        104     574    207   
executive         808        436         91        232     951    287   
healthcare        577        323         89        213     722    200   
homemaker          92         41          9         25      93     25   
lawyer            283        142         35         83     460    107   
librarian         988        536        159        290    1576    370   
marketing         477        246         79        163     535    161   
none              301        160         43         91     255     74   
other            2672       1431        421        881    3165    828   
programmer       2322       1240        312        583    2418    666   
retired           278        160         35        110     512    137   
salesman          222         99         15         38     314     79   
scientist         566        285         61        108     535    177   
student          6398       3358        914       1759    6958   1836   
technician       1079        558        132        221    1056    275   
writer           1292        674        176        367    1574    498   

               Documentary  Drama  Fantasy  Film-Noir  Horror  Musical  \
occupation                                                               
administrator           48   3099       94        147     391      354   
artist                  34    957       24         64     102      157   
doctor                  13    238        7          9      18       27   
educator               107   4281      112        202     421      481   
engineer                51   3153      125        145     414      441   
entertainment           23    804       25         47     192       68   
executive               26   1407       50         58     155      131   
healthcare              45   1297       38         50     120      126   
homemaker                2    104        3          4       9       15   
lawyer                  10    543       14         36      60       72   
librarian               48   2557       55         96     226      238   
marketing               12    824       20         40      86      114   
none                     1    268       12         12      56       43   
other                   65   4307      144        146     535      566   
programmer              51   2800      117        141     389      385   
retired                 16    727       22         43      66       93   
salesman                 2    317        9         14      49       27   
scientist               17    932       20         39      70       91   
student                117   7777      352        269    1340     1078   
technician              24   1295       38         46     164      175   
writer                  46   2208       71        125     454      272   

               Mystery  Romance  Sci-Fi  Thriller   War  Western  
occupation                                                        
administrator      385     1539     942      1570   786      168  
artist             164      462     308       476   225       29  
doctor              27      133      56       117    44        5  
educator           532     2006     986      1767  1005      189  
engineer           363     1467    1221      1712   858      180  
entertainment      132      342     289       554   174       27  
executive          203      592     391       811   312       49  
healthcare         140      4

In [74]:
df_merge_all.groupby('occupation').sum().loc[:,'Action':'Western'].loc['scientist']

Action         566
Adventure      285
Animation       61
Childrens      108
Comedy         535
Crime          177
Documentary     17
Drama          932
Fantasy         20
Film-Noir       39
Horror          70
Musical         91
Mystery        122
Romance        440
Sci-Fi         318
Thriller       469
War            255
Western         24
Name: scientist, dtype: int64

**Observation:**   Engineers mostly watch movies of Drama, Comedy, and Action genres.

You can try finding the most common genre among different professionals.

### We have just learned how to use the pandas cut() function to group a numerical column in the data into different categories and use agg() to use multiple aggregate functions at once with groupby(). These are important techniques that allow us to create informative groups from numerical columns and understand multiple statistics of subsets of our data at once.